// @aws-cdk/core/test/private/test.tree-metadata.js
"use strict";
const fs = require("fs");
const path = require("path");
const index_1 = require("../../lib/index");
class MyResource extends index_1.Resource {
    constructor(scope, id) {
        super(scope, id);
    }
}
function readJson(outdir, file) {
    return JSON.parse(fs.readFileSync(path.join(outdir, file), 'utf-8'));
}
module.exports = {
    'tree metadata is generated as expected'(test) {
        const app = new index_1.App();
        const stack = new index_1.Stack(app, 'mystack');
        new index_1.Construct(stack, 'group1');
        const group2 = new index_1.Construct(stack, 'group2');
        new MyResource(group2, 'resource3');
        const assembly = app.synth();
        const treeArtifact = assembly.tree();
        test.ok(treeArtifact);
        test.deepEqual(readJson(assembly.directory, treeArtifact.file), {
            version: 'tree-0.1',
            tree: {
                id: 'App',
                path: '',
                children: [
                    {
                        id: 'Tree',
                        path: 'Tree'
                    },
                    {
                        id: 'mystack',
                        path: 'mystack',
                        children: [
                            {
                                id: 'group1',
                                path: 'mystack/group1'
                            },
                            {
                                id: 'group2',
                                path: 'mystack/group2',
                                children: [
                                    { id: 'resource3', path: 'mystack/group2/resource3' }
                                ]
                            }
                        ]
                    },
                ]
            }
        });
        test.done();
    },
};
